File: Neetcode/ConstructBTfromTraversals.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        h = TreeNode(preorder[0])
        head = h
        nodeindex = dict()
        for i in range(len(inorder)):
            nodeindex[inorder[i]] = i
        headindex = nodeindex[preorder[0]]
        i = 1
        while i < len(preorder):
            if nodeindex[preorder[i]] < headindex:
                if head.left is None:
                    head.left = TreeNode(preorder[i])
                    i += 1
                else:
                    head = head.left
                    headindex = nodeindex[head.val]
                    continue
            else:
                if head.right is None:
                    head.right = TreeNode(preorder[i])
                    i += 1
                else:
                    head = head.right
                    headindex = nodeindex[head.val]
                    continue
            head = h
            headindex = nodeindex[head.val]
        return h

File: Neetcode/ConstructBTfromTraversals.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        unordered_map<int, int> dict;
        for(int i=0; i<inorder.size(); i++){
            dict[inorder[i]] = i;
        }
        TreeNode* root = nullptr;
        int index = -1;
        for(int i=0; i<preorder.size(); i++){
            if(i==0){
                root = new TreeNode(preorder[i]);
                index = dict[preorder[i]];
            }
            else{
                int j = index;
                TreeNode* node = root;
                while(true){
                    if(dict[preorder[i]] < j){
                        if(node->left == nullptr){
                            node->left = new TreeNode(preorder[i]);
                            break;
                        }
                        else{
                            node = node->left;
                            j = dict[node->val];
                            continue;
                        }
                    }
                    else{
                        if(node->right == nullptr){
                            node->right = new TreeNode(preorder[i]);
                            break;
                        }
                        else{
                            node = node->right;
                            j = dict[node->val];
                            continue;
                        }
                    }
                }
            }
        }
        return root;
    }
};
